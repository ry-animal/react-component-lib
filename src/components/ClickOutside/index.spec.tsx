import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "jest-styled-components";
import { ClickOutside } from ".";

describe("<ClickOutside />", () => {
  describe("rendering", () => {
    test("Default rendering matches snapshot", () => {
      expect(
        render(
          <ClickOutside onClick={jest.fn()}>Inner Content</ClickOutside>
        ).asFragment()
      ).toMatchSnapshot();
    });
  });

  describe("event handling", () => {
    test("Clicking inside does not fire outside click", () => {
      const outsideClick = jest.fn();
      const insideClick = jest.fn();
      render(
        <ClickOutside onClick={outsideClick}>
          <button onClick={insideClick}>Click</button>
        </ClickOutside>
      );

      userEvent.click(screen.getByRole("button"));
      expect(insideClick).toBeCalled();
      expect(outsideClick).not.toBeCalled();
    });

    test("Clicking outide fires outside click", () => {
      const outsideClick = jest.fn();
      const insideClick = jest.fn();
      render(
        <ClickOutside onClick={outsideClick}>
          <button onClick={insideClick}>Click</button>
        </ClickOutside>
      );

      const outerDiv = screen.getByRole("button").closest("div");
      if (outerDiv) {
        userEvent.click(outerDiv);
      }
      expect(insideClick).not.toBeCalled();
      expect(outsideClick).toBeCalled();
    });

    test("Clicking 'ESC' fires outside click", () => {
      const onClick = jest.fn();
      const { container } = render(
        <ClickOutside onClick={onClick}>Inner Content</ClickOutside>
      );

      fireEvent.keyDown(container, {
        key: "Escape",
      });

      expect(onClick).toBeCalled();

      fireEvent.keyDown(container, {
        key: "Esc",
      });

      expect(onClick).toBeCalled();
    });
  });
});
