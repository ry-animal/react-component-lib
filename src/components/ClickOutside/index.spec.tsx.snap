// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClickOutside /> rendering Default rendering matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  height: 100%;
  position: relative;
}

<div
    class="c0"
  >
    Inner Content
  </div>
</DocumentFragment>
`;
